### ISSUE TITLE
fd closing for scriptlet execution seems to be done inefficiently

### ISSUE THREAD

**Issue by keszybz** [rpm-software-management/rpm#2081]
I was debugging some scriptlet, and noticed the following from `sudo strace dnf …`:

```
231692 write(1, "\r  Running scriptlet: rpmexec-0-1.fc37.noarch                                                              
               1/1", 125) = 125
231692 write(1, " \n", 2)               = 2
231692 openat(AT_FDCWD, ".", O_RDONLY)  = 24
231692 umask(000)                       = 022
231692 umask(022)                       = 000
231692 getpid()                         = 231692
231692 chdir("/")                       = 0
231692 clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f2c19b9ca10) = 231697
231692 wait4(231697,  <unfinished ...>
231697 set_robust_list(0x7f2c19b9ca20, 24) = 0
231697 openat(AT_FDCWD, "/dev/null", O_WRONLY) = 25
231697 dup2(25, 2)                      = 2
231697 close(25)                        = 0
231697 openat(AT_FDCWD, "/proc/self/fd", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = -1 ENOENT (No such file or directory)
231697 prlimit64(0, RLIMIT_NOFILE, NULL, {rlim_cur=1024, rlim_max=512*1024}) = 0
231697 fcntl(3, F_GETFD)                = 0x1 (flags FD_CLOEXEC)
231697 fcntl(4, F_GETFD)                = 0x1 (flags FD_CLOEXEC)
231697 fcntl(5, F_GETFD)                = 0
231697 fcntl(5, F_SETFD, FD_CLOEXEC)    = 0
231697 fcntl(6, F_GETFD)                = 0x1 (flags FD_CLOEXEC)
231697 fcntl(7, F_GETFD)                = 0
231697 fcntl(7, F_SETFD, FD_CLOEXEC)    = 0
231697 fcntl(8, F_GETFD)                = 0x1 (flags FD_CLOEXEC)
231697 fcntl(9, F_GETFD)                = 0x1 (flags FD_CLOEXEC)
231697 fcntl(10, F_GETFD)               = 0x1 (flags FD_CLOEXEC)
231697 fcntl(11, F_GETFD)               = 0x1 (flags FD_CLOEXEC)
231697 fcntl(12, F_GETFD)               = 0x1 (flags FD_CLOEXEC)
231697 fcntl(13, F_GETFD)               = 0x1 (flags FD_CLOEXEC)
231697 fcntl(14, F_GETFD)               = 0x1 (flags FD_CLOEXEC)
231697 fcntl(15, F_GETFD)               = 0
231697 fcntl(15, F_SETFD, FD_CLOEXEC)   = 0
231697 fcntl(16, F_GETFD)               = 0x1 (flags FD_CLOEXEC)
231697 fcntl(17, F_GETFD)               = 0x1 (flags FD_CLOEXEC)
231697 fcntl(18, F_GETFD)               = 0x1 (flags FD_CLOEXEC)
231697 fcntl(19, F_GETFD)               = 0
231697 fcntl(19, F_SETFD, FD_CLOEXEC)   = 0
231697 fcntl(20, F_GETFD)               = 0
231697 fcntl(20, F_SETFD, FD_CLOEXEC)   = 0
231697 fcntl(21, F_GETFD)               = 0x1 (flags FD_CLOEXEC)
231697 fcntl(22, F_GETFD)               = 0x1 (flags FD_CLOEXEC)
231697 fcntl(23, F_GETFD)               = 0x1 (flags FD_CLOEXEC)
231697 fcntl(24, F_GETFD)               = 0
231697 fcntl(24, F_SETFD, FD_CLOEXEC)   = 0
231697 fcntl(25, F_GETFD)               = -1 EBADF (Bad file descriptor)
...
231697 fcntl(524285, F_GETFD)           = -1 EBADF (Bad file descriptor)
231697 fcntl(524286, F_GETFD)           = -1 EBADF (Bad file descriptor)
231697 fcntl(524287, F_GETFD)           = -1 EBADF (Bad file descriptor)
```

524287 is `ulimit -H -n`. I assume that rpm wants to make sure that all fds are closed, but `/proc` is not mounted for scriptlets, so the code can't enumerate fds by listing `/proc/self/fd`, so it falls back to a loop over all possible fds. For some reason it uses the hard limit, even though it should be using the soft one, which is usually much lower (1024 by default for compat with select(2)).

This is painful, doing ~500k system calls for each scriptlet adds up.

<details>
Reproducer:
```rpm
# rpmexec.spec
Name:           rpmexec
Version:        0
Release:        %autorelease
Summary:        None

License:        None
URL:            None
BuildArch:      noarch

%description
%{summary}.

%prep
echo 'asdf' >license

%post -p <lua>
local pid = posix.fork()
if pid == 0 then
  posix.redirect2null(2)
  posix.exec("echo", "Hello, World!")
elseif pid > 0 then
  posix.wait(pid)
end

%files
%license license

%changelog
%autochangelog
```
`fedpkg --release=rawhide local`
`sudo strace -s600 -f -o /tmp/log dnf install --repo=rawhide --releasever=rawhide --installroot=/var/tmp/inst1 noarch/rpmexec-0-1.fc37.noarch.rpm -y`
</details>

**Comment by pmatilai** [comment]
> /proc is not mounted for scriptlets

What environment is that? If /proc is not mounted then a whole lot of things will be quite broken, and I don't mean just rpm.

> For some reason it uses the hard limit, even though it should be using the soft one

From commit 307e28b4cb08b05bc044482058eeebc9f59bb9a9:

    In case /proc is not available to get the actual list of opened fds,
    we fall back to iterating through the list of all possible fds.
    
    It is possible that during the course of the program execution the limit
    on number of open file descriptors might be lowered, so using the
    current limit, as returned by sysconf(_SC_OPEN_MAX), might omit some
    fds. Therefore, it is better to use rlim_max from the structure
    filled in by gertlimit(RLIMIT_NOFILE) to make sure we're checking
    all fds.
    
    This slows down the function, but only in the case /proc is not
    available, which should be rare in practice.

**Comment by keszybz** [comment]
That's not how things work in practice. We have bucketloads of rpms with scriptlets, and people use either `dnf --installroot` or something equivalent to populate images. In fact, I'd say that nowadays this is more common than ever, and things will only continue to move in this direction, as installation of rpms on a live system will be less and less common. Dunno, I guess this isn't clearly documented anywhere, but the packaging practice has always been to have scriptlets work in the minimal environment provided by `rpm`.

**Comment by pmatilai** [comment]
I don't think we disagree at all, just looking at the same thing from different perspectives. From what I've seen, the "core" stuff (in Fedora) usually does work with plain --root whereas the higher up you go, the more things explode every which way if /proc is not mounted. So people use have these wrappers like mock which set up the environment to resemble more something workable. And because stuff like mock exists nobody has ever so much as *asked* rpm to set up anything more.

**Comment by keszybz** [comment]
FWIW, we (speaking loosely here about myself and other people I worked with) didn't ask for rpm to do something more because it was always easy enough to get things working without `/proc`.
But the corollary here seems to be that rpm itself is not ready for running efficiently in this environment…

> From what I've seen, the "core" stuff (in Fedora) usually does work with plain --root whereas the higher up you go, the more things explode every which way if /proc is not mounted.

Hmm, maybe. This probably varies a lot between cases, but from what I've seen, it could be said that even the opposite is true: "low-level" rpms (in particular systemd, but also other things that are often installed early in the transaction) need to do twiddling and setup things up, so they encounter problems with missing tools and/or an environment that hasn't been set up yet. But various "high-level" rpms do much less custom scriptlets. In particular, many scriptlets that those packages use are some macroified stuff provided by the low-level rpms…

**Comment by pmatilai** [comment]
Well, rpm does not set up any other mounts either. The general expectation simply is that somebody else has done that before running a transaction. Actually documenting this status quo someplace wouldn't be a bad thing...

**Comment by cgwalters** [comment]
rpm-ostree hard requires a container runtime (currently `bwrap`) and uses it unconditionally for scripts for this reason, among others.  We don't support a bare `--installroot` type flow. It never made any sense to me at all to have something on top like `mock` handle this.

**Comment by DemiMarie** [comment]
> rpm-ostree hard requires a container runtime (currently `bwrap`) and uses it unconditionally for scripts for this reason, among others. We don't support a bare `--installroot` type flow. It never made any sense to me at all to have something on top like `mock` handle this.

I strongly support this approach.  RPM currently calls `chroot` which is a really bad thing for a library to do.  RPM should be accessing the install root from within a separate process.
